### ISSUE TITLE
Repository with example implementation of sentiment classification with Flair

### ISSUE THREAD

**Issue by krzysztoffiok** [flairNLP/flair#1565]
Hi,

i've just published [a repository](https://github.com/krzysztoffiok/twitter_sentiment) that uses Flair for implementation of sentiment classification. The repository contains also an example of fine-tuning selected transformer model which is a recently added feature to master branch of Flair.
The repository compares various models for feature extraction (Term Frequency, LIWC, pooled features from FastText and RoBERTa, features created by biLSTM over FastText and RoBERTa, Universal Sentence Encoder, CLS output from RoBERTa) and demonstrates machine learning implementation on top. In the last section it also shows how SHAP XAI explanations can work in such setup.

In the published code one can use already extracted embeddings and play only with the machine learning PART II or carry out also model training in PART I.

Works regarding publishing this as a research paper are ongoing.

Thanks again to all developers of Flair for this great framework!

p.s.

One observation is that when i prepared a demo notebook on Google Colaboratory i found that "scalar_mix=True" made colab crash with an error reporting problems with choice of torch.device (it would work with CPU but that is incredibly slow).

**Comment by alanakbik** [comment]
@krzysztoffiok awesome, thanks for sharing this! Is there a table with an overview of all evaluation results? Which approach worked the best?

**Comment by krzysztoffiok** [comment]
I haven't published the final results yet as we still experiment with ML parameters. If we consider 2 ML classifiers:
1) RandomForestClassifier(n_estimators=250, max_depth=7, min_samples_split=2, min_samples_leaf=1, max_features='auto', n_jobs=-1, random_state=2020),
2) xgb.XGBClassifier(objective='multi:softprob', n_jobs=24, learning_rate=0.03,
                                 max_depth=10, subsample=0.7, colsample_bytree=0.6, random_state=2020, n_estimators=250) 

then for the embeddings provided in the "release" section of the above mentioned repository the results are more or less this:


  |   |   |   |   | Flair | Flair | Flair | Flair | Flair
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   | Pooled | Pooled | biLSTM | biLSTM | CLS
Metric | Model | Term Frequency | LIWC | USE | FastText | RoBERTa | FastText | RoBERTa | RoBERTa
MCC | RF | 0.356 | 0.425 | 0.461 | 0.421 | 0.449 | 0.524 | 0.594 | 0.609
MCC | XGB | 0.465 | 0.466 | 0.537 | 0.48 | 0.541 | 0.527 | 0.586 | 0.613
F1 | RF | 0.498 | 0.596 | 0.582 | 0.557 | 0.579 | 0.665 | 0.721 | 0.732
F1 | XGB | 0.615 | 0.628 | 0.672 | 0.63 | 0.678 | 0.673 | 0.719 | 0.736

So if one is not interested in training a Language Model the best choice is to take Universal Sentence Encoder (USE) embeddings or pooled RoBERTa embeddings. If however one wants the best possible performance, fine-tuning of RoBERTa and using its CLS output will provide it. If explainability of model predictions is required, unfortunately to use SHAP explanations one has to stick to Term Frequency or LIWC models. For other models explanations can also be computed, but they don't help much since it is unknown what the given features represent.

All these results are from a 5-scale sentiment analysis carried out for 5000 tweets.

**Comment by alanakbik** [comment]
Thanks for sharing your results! Looks like RoBERTa cannot be beaten ;) 

What is the score you get during the fine-tuning of RoBERTa CLS token? Does fitting XGB on top improve results, or could one just use RoBERTa fine-tuning as-is?

**Comment by krzysztoffiok** [comment]
I will try to investigate that, i'm not sure right now. Only it will take me some time as i'm unable to work on that right away.

**Comment by krzysztoffiok** [comment]
OK, sorry for the delay @alanakbik. Here is the comparison i was able to make without changing metrics computed by Flair. Flair metrics values are taken from the training.log after testing with the best_model.pt for each CV fold. My metrics were computed with sklearn.metrics.f1_score. Here is a .zip with training logs and files used for computing of below table.
[_training_logs.zip](https://github.com/flairNLP/flair/files/4579416/_training_logs.zip)


  | Flair Neural Classifier |   | XGBoost n_estimates=250 |  
-- | -- | -- | -- | --
  | F1 |   | F1 |  
CV fold | micro | macro | micro | macro
0 | 0.732 | 0.431 | 0.76 | 0.62
1 | 0.756 | 0.566 | 0.748 | 0.606
2 | 0.735 | 0.544 | 0.754 | 0.617
3 | 0.728 | 0.54 | 0.748 | 0.594
4 | 0.726 | 0.578 | 0.747 | 0.589
average | 0.735 | 0.532 | 0.751 | 0.605
std | 0.011 | 0.052 | 0.005 | 0.012


The difference seems to be very significant in F1 "macro" score and less significant in "micro". The Flair classifier "macro" score in fold 0 is low because the model didn't predict correctly any instance in class 0 and 4 (see attached logs for details, my explanation is that the data sample is small = only 5000 instances).

Notice on score computation: In case of this comparison all scores were computed per fold and averaged after, which gives slightly different results from other possibility which is: aggregating all predictions and computing scores once for the whole experiment (in previously published results this was the way of computing scores and they were "weighted" so different than here). To get an idea of how this affects the comparison: F1 micro average if computed per fold and averaged was 0.751 and the latter option results in 0.747; F1 macro: 0.605 and the latter option 0.589.

Summary: if there are any real differences between XGBoost and Neural Classifier built in Flair, they are small. One should carry out a similar comparison to this here, yet on a larger data sample.

**Comment by djstrong** [comment]
I think you are not optimizing for F1 but for accuracy. Beside that, your model is overfitting.

**Comment by krzysztoffiok** [comment]
@djstrong thank you for feedback, i understand you mean that in these 4 epochs loss on train split decreases fast, but the score (accuracy) on validation set decreases as well? any ideas how to avoid this? make the learning rate smaller?

also, if i wish to optimize F1 instead of accuracy - can i change it? in trainer parameters perhaps?

**Comment by alanakbik** [comment]
@djstrong thanks for sharing your results. So the XGBoost is fitted on top of the result of the Flair experiment, right? i.e. you use the fine-tuned transformer to produce embeddings and then train XGBoost on top? 

Since performance is best in the first epoch in the logs, a lower learning rate makes sense. In parallel, you could try modifying the mini_batch_size. It's a small corpus so a lower one (maybe 16) could be tried in combination with a lower learning rate. You could also try a larger mini_batch_size (64) to see if this prevents overfitting too quickly.

**Comment by krzysztoffiok** [comment]
@alanakbik yes, XGBoost trained on top of RoBERTa CLS embeddings from a Flair fine-tuned model are compared with classifier available in Flair.

Lowering learning rate from 3e-5 to 3e-6 actually didn't work well with classifier from Flair. Just finished fine-tuning and it gave poorer results (even though validation accuracy was higher than previously, logs attached 
[_tlogs2.zip](https://github.com/flairNLP/flair/files/4580661/_tlogs2.zip)
)

  | Flair Neural Classifier |  
-- | -- | --
  | F1 |  
CV fold | micro | macro
0 | 0.721 | 0.566
1 | 0.716 | 0.491
2 | 0.741 | 0.468
3 | 0.726 | 0.482
4 | 0.724 | 0.58
average | 0.726 | 0.517
std | 0.008 | 0.046

But thanks for other suggestions, i'll give it a try.

**Comment by alanakbik** [comment]
Interesting, thanks for sharing this. If even without any parameter tuning XGBoost already improves the scores, perhaps this should always be done for text classification?

**Comment by krzysztoffiok** [comment]
Well, i think this can be regarded as a classic machine learning problem where you have some features and try to fit a classifier. Maybe possibly better performance of XGBoost can be attributed to the small corpus size? My choice of approach, to use a classifier external to Flair, was determined by the will to compare several approaches for creating embeddings, not only these available in Flair, and use the same classification method.

**Comment by alanakbik** [comment]
Yes, perhaps it would be good to offer a better ScikitLearn integration to Flair in the future so that such experiments can be done. Also, we should probably add USE.
